### PR TITLE
[dynamic_links]: fix initialization process

### PR DESCRIPTION
**Pull request by jacek-marchwicki**
Description:

This is my first contribution to this project so I'm sorry if I made this in improper way.

## Description

Fix race conditions related to the dynamic links initialization process.

1. During installation flow `- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation` might be called by OS after `FirebaseDynamicLinks.instance.getInitialLink()` is called. In such a case the app won't retrieve the deep-link.
2. If `FirebaseDynamicLinks.instance.getInitialLink()` is called while processing `[FIRDynamicLinks dynamicLinks]  openURL:(NSURL *)url handleUniversalLink`, the url is lost.
3. Documentation fix: `FirebaseDynamicLinks.instance.getInitialLink()` call should be called after `FirebaseDynamicLinks.instance.onLink()` so url's aren't lost between those two calls.

More info in the comments in the PR:

## Related Issues

* https://github.com/FirebaseExtended/flutterfire/issues/100
* https://github.com/FirebaseExtended/flutterfire/issues/1847
* https://github.com/FirebaseExtended/flutterfire/issues/1861

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]). - **I haven't wrote any integration tests for the iOS stuff because i'd be a little complex.**
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m [hunk 1]
```diff
@@ -118,63 +118,57 @@ - (NSMutableDictionary *)getInitialLink {
...
+            _initialLink = dynamicLink;
+        }
+    }
 }
 
 - (BOOL)checkForDynamicLink:(NSURL *)url {
   FIRDynamicLink *dynamicLink = [[FIRDynamicLinks dynamicLinks] dynamicLinkFromCustomSchemeURL:url];
   if (dynamicLink) {
-    if (dynamicLink.url) _initialLink = dynamicLink;
+    [self onDeepLinkResult: dynamicLink error: nil];
```
Comment (jacek-marchwicki) [inline packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m:144]: info: when the `checkForDynamicLink` method is executed, the `getInitialLink()` method had been already called. In such a case the app won't read `_initialLink`.
After the change, the link is delivered to `onLink(onSuccess: xyz)` callback

File: packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m [hunk 2]
```diff
@@ -118,63 +118,57 @@ - (NSMutableDictionary *)getInitialLink {
...
 - (BOOL)application:(UIApplication *)application
-    continueUserActivity:(NSUserActivity *)userActivity
-      restorationHandler:(void (^)(NSArray *))restorationHandler {
-  if (_initiated) {
-    return [self onLink:userActivity];
-  }
-  return [self onInitialLink:userActivity];
+continueUserActivity:(NSUserActivity *)userActivity
+ restorationHandler:(void (^)(NSArray *))restorationHandler {
+    BOOL handled = [[FIRDynamicLinks dynamicLinks]
```
Comment (jacek-marchwicki) [inline packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m:166]: info: previously `_initiated` were checked before calling async function `[[FIRDynamicLinks dynamicLinks] handleUniversalLink:userActivity`. So previously if `getInitialLink` is called after the method is called but before it returns data. The link is lost.
After the change, the `_initiated` is checked after the callback returns.

File: packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m [hunk 3]
```diff
@@ -118,63 +118,57 @@ - (NSMutableDictionary *)getInitialLink {
   return getDictionaryFromDynamicLink(_initialLink);
 }
 
-- (BOOL)application:(UIApplication *)application
-            openURL:(NSURL *)url
-            options:(NSDictionary<UIApplicationOpenURLOptionsKey, id> *)options {
-  return [self checkForDynamicLink:url];
-}
```
Comment (jacek-marchwicki) [inline packages/firebase_dynamic_links/ios/Classes/FLTFirebaseDynamicLinksPlugin.m:125]: info: nothing has changed, the method is moved lower where the rest of `-application: xxx` method are placed.

**Comment by jacek-sf** [comment]
This PR is ready for review by I can't change it from draft.

**Comment by jacek-marchwicki** [comment]
Hi @bparrishMines!

Great that you taking care of this PR. Is it something that I can do for you related to this PR?

**Comment by kroikie** [comment]
@jacek-marchwicki could you please bump the version number and update the changelog?

**Comment by jacek-marchwicki** [comment]
@kroikie sorry it took me a while, but finally I did it!

**Comment by sgehrman** [comment]
awesome fix!  I spent 2 days trying to figure out why my links were acting weird.

**Pull request merged**